src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/Data/PathParser.cs

```diff
@@ -189,7 +189,7 @@ void AddProperty()
             while (_index < _n && _path[_index] == '.')
                 ++_index;
 
-            while (_index < _n && (level > 0 || SpecialChars.IndexOf(_path[_index]) < 0))
+            while (_index < _n && (level > 0 || !IsSpecialChar(_path[_index])))
```

Comment:
You can get even more performance if the bound-checks* are eleminated.
`_n` is the length of the path, so use this information for the JIT.

```suggestion
            string path = _path;
            int index = _index;
            while ((uint)index < (uint)path.Length && (level > 0 || !IsSpecialChar(path[index])))
```
Removes the bound-checks.
The same could be done at other places -- and get rid of the (unnecessary) `_n` field (makes the parser 4-bytes smaller, not much but a little bit and it harms perf as the JIT can't remove bound-checks).

\* the JIT needs to emit these bound-checks in order to prove the access to the array is within bound. Only if the JIT is sure the access won't be outside, these can be removed.

Reply:
I already tried removing the bound-checks but I didn't see any perf gain. Though I didn't use the uint cast. I will try to benchmark it again with the cast and I can open another PR with your suggestions if you'd like.

Reply:
The `uint`-cast is essential for the JIT, as any negative value (which is allowed by `int`) will be > int.MaxValue and so for sure outside the bounds. Then when the JIT sees that index is non-negative and < length it's OK to elide the bound-check.

Though the JIT could be improved in that area (and it is).